

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>pp &mdash; STORM  documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="_static/jupyter-sphinx.css?v=572af1d6" />
      <link rel="stylesheet" type="text/css" href="_static/thebelab.css" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=5929fcd5"></script>
      <script src="_static/doctools.js?v=9a2dae69"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
      <script src="_static/thebelab-helper.js"></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
      <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="gt" href="gt.html" />
    <link rel="prev" title="VisiumReader" href="VisiumReader.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            STORM
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Content</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="HDReader.html">HDReader</a></li>
<li class="toctree-l1"><a class="reference internal" href="VisiumReader.html">VisiumReader</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">pp</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#storm.pp.Cal_CMatrix"><code class="docutils literal notranslate"><span class="pre">Cal_CMatrix</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#storm.pp.Cal_CMatrix.get_cmtx"><code class="docutils literal notranslate"><span class="pre">Cal_CMatrix.get_cmtx()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#storm.pp.Cal_CMatrix.readFile"><code class="docutils literal notranslate"><span class="pre">Cal_CMatrix.readFile()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#storm.pp.HDPreprocesser"><code class="docutils literal notranslate"><span class="pre">HDPreprocesser</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#storm.pp.HDPreprocesser.process_grid"><code class="docutils literal notranslate"><span class="pre">HDPreprocesser.process_grid()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#storm.pp.HDPreprocesser.process_image"><code class="docutils literal notranslate"><span class="pre">HDPreprocesser.process_image()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#storm.pp.HDPreprocesser.process_matrix"><code class="docutils literal notranslate"><span class="pre">HDPreprocesser.process_matrix()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#storm.pp.HDPreprocesser.saveFile"><code class="docutils literal notranslate"><span class="pre">HDPreprocesser.saveFile()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#storm.pp.VisiumPreprocesser"><code class="docutils literal notranslate"><span class="pre">VisiumPreprocesser</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#storm.pp.VisiumPreprocesser.cal_cor"><code class="docutils literal notranslate"><span class="pre">VisiumPreprocesser.cal_cor()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#storm.pp.VisiumPreprocesser.concat_adata"><code class="docutils literal notranslate"><span class="pre">VisiumPreprocesser.concat_adata()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#storm.pp.VisiumPreprocesser.crop_img"><code class="docutils literal notranslate"><span class="pre">VisiumPreprocesser.crop_img()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#storm.pp.VisiumPreprocesser.expand_barcode"><code class="docutils literal notranslate"><span class="pre">VisiumPreprocesser.expand_barcode()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#storm.pp.VisiumPreprocesser.find_avg_grid"><code class="docutils literal notranslate"><span class="pre">VisiumPreprocesser.find_avg_grid()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#storm.pp.VisiumPreprocesser.generate_grid"><code class="docutils literal notranslate"><span class="pre">VisiumPreprocesser.generate_grid()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#storm.pp.VisiumPreprocesser.insert_grid"><code class="docutils literal notranslate"><span class="pre">VisiumPreprocesser.insert_grid()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#storm.pp.VisiumPreprocesser.map_tissue"><code class="docutils literal notranslate"><span class="pre">VisiumPreprocesser.map_tissue()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#storm.pp.VisiumPreprocesser.mark_bc_label"><code class="docutils literal notranslate"><span class="pre">VisiumPreprocesser.mark_bc_label()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#storm.pp.VisiumPreprocesser.mark_tissue_grid"><code class="docutils literal notranslate"><span class="pre">VisiumPreprocesser.mark_tissue_grid()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#storm.pp.VisiumPreprocesser.mark_tissue_label"><code class="docutils literal notranslate"><span class="pre">VisiumPreprocesser.mark_tissue_label()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#storm.pp.VisiumPreprocesser.process_adata"><code class="docutils literal notranslate"><span class="pre">VisiumPreprocesser.process_adata()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#storm.pp.VisiumPreprocesser.process_img"><code class="docutils literal notranslate"><span class="pre">VisiumPreprocesser.process_img()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#storm.pp.VisiumPreprocesser.process_tpl"><code class="docutils literal notranslate"><span class="pre">VisiumPreprocesser.process_tpl()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#storm.pp.VisiumPreprocesser.round_spot"><code class="docutils literal notranslate"><span class="pre">VisiumPreprocesser.round_spot()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#storm.pp.VisiumPreprocesser.save"><code class="docutils literal notranslate"><span class="pre">VisiumPreprocesser.save()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#storm.pp.downsample_svs"><code class="docutils literal notranslate"><span class="pre">downsample_svs()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#storm.pp.exc_he"><code class="docutils literal notranslate"><span class="pre">exc_he()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#storm.pp.exc_tissue"><code class="docutils literal notranslate"><span class="pre">exc_tissue()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#storm.pp.extract_tiles"><code class="docutils literal notranslate"><span class="pre">extract_tiles()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#storm.pp.process_folder"><code class="docutils literal notranslate"><span class="pre">process_folder()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#storm.pp.white_balance_using_white_point"><code class="docutils literal notranslate"><span class="pre">white_balance_using_white_point()</span></code></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="gt.html">gt</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">STORM</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">pp</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/pp.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-storm.pp">
<span id="pp"></span><h1>pp<a class="headerlink" href="#module-storm.pp" title="Link to this heading"></a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="storm.pp.Cal_CMatrix">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">storm.pp.</span></span><span class="sig-name descname"><span class="pre">Cal_CMatrix</span></span><a class="headerlink" href="#storm.pp.Cal_CMatrix" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A class for calculating color matrices.</p>
<dl class="py method">
<dt class="sig sig-object py" id="storm.pp.Cal_CMatrix.get_cmtx">
<span class="sig-name descname"><span class="pre">get_cmtx</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'vahadane'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.Cal_CMatrix.get_cmtx" title="Link to this definition"></a></dt>
<dd><p>Get the calculated color matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>method</strong> (<em>str</em><em>, </em><em>optional</em>) – The method for stain matrix calculation. should be “vahadane” or “macenko”.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The calculated stain color matrix.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="storm.pp.Cal_CMatrix.readFile">
<span class="sig-name descname"><span class="pre">readFile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">raw_img_path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.Cal_CMatrix.readFile" title="Link to this definition"></a></dt>
<dd><p>Read an image file and store it in the class instance.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>raw_img_path</strong> (<em>str</em>) – The path to the raw image file.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="storm.pp.HDPreprocesser">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">storm.pp.</span></span><span class="sig-name descname"><span class="pre">HDPreprocesser</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">Reader</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">token_path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.HDPreprocesser" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A class for processing High - Definition (HD) data.</p>
<p>This class provides methods to process the HD data read by an HDReader instance.
It can process the grid, image, tissue, and matrix data, and save the processed data to files.</p>
<dl class="py method">
<dt class="sig sig-object py" id="storm.pp.HDPreprocesser.process_grid">
<span class="sig-name descname"><span class="pre">process_grid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">patch_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dst_res</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_res</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.HDPreprocesser.process_grid" title="Link to this definition"></a></dt>
<dd><p>Generate an image and tissue grid based on the tissue position and spot - to - microscope information.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>patch_size</strong> (<em>int</em>) – The size of the patches used in generating the image grid.</p></li>
<li><p><strong>dst_res</strong> (<em>int</em>) – The destination resolution for the image grid.</p></li>
<li><p><strong>bin_res</strong> (<em>int</em><em>, </em><em>optional</em>) – The binning resolution. Defaults to 2.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="storm.pp.HDPreprocesser.process_image">
<span class="sig-name descname"><span class="pre">process_image</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.HDPreprocesser.process_image" title="Link to this definition"></a></dt>
<dd><p>Process the image based on the image grid information.</p>
<p>This method first converts specific columns of the image grid to NumPy arrays,
then calculates the pixels using these arrays and the original image,
and finally post - processes the calculated image.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The processed image is stored in the ‘nwimg’ attribute of the class.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="storm.pp.HDPreprocesser.process_matrix">
<span class="sig-name descname"><span class="pre">process_matrix</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.HDPreprocesser.process_matrix" title="Link to this definition"></a></dt>
<dd><p>Process the matrix data and update the AnnData object.</p>
<p>This method deletes the image grid, original image, and tissue position information
to free up memory, then renews the AnnData object based on the tissue grid,
sets the gene token, and binarizes the matrix data.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The processed AnnData object is stored in the ‘adata’ attribute of the class.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="storm.pp.HDPreprocesser.saveFile">
<span class="sig-name descname"><span class="pre">saveFile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">imgrid_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fhe_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">h5ad_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.HDPreprocesser.saveFile" title="Link to this definition"></a></dt>
<dd><p>Save the processed data to files.</p>
<p>This method can save the image grid, processed H&amp;E image, and AnnData object to files
if the corresponding file paths are provided.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>imgrid_path</strong> (<em>str</em><em>, </em><em>optional</em>) – The path to save the image grid as a Parquet file. Defaults to None.</p></li>
<li><p><strong>fhe_path</strong> (<em>str</em><em>, </em><em>optional</em>) – The path to save the processed H&amp;E image. Defaults to None.</p></li>
<li><p><strong>h5ad_path</strong> (<em>str</em><em>, </em><em>optional</em>) – The path to save the AnnData object. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="storm.pp.VisiumPreprocesser">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">storm.pp.</span></span><span class="sig-name descname"><span class="pre">VisiumPreprocesser</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">Reader</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">patch_size</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.VisiumPreprocesser" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A class for processing Visium spatial data.</p>
<p>This class takes a file reader object and patch size as inputs,
and provides a series of methods to process tissue position data,
H&amp;E images, and gene expression data. The processed data can be saved to files.</p>
<dl class="py method">
<dt class="sig sig-object py" id="storm.pp.VisiumPreprocesser.cal_cor">
<span class="sig-name descname"><span class="pre">cal_cor</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.VisiumPreprocesser.cal_cor" title="Link to this definition"></a></dt>
<dd><p>Generate tissue grid by tissue bounding box.</p>
<p>This method finds the bounding box of the tissue positions within the tissue.
It ensures that the dimensions of the bounding box are divisible by the patch size.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><dl class="simple">
<dt>A tuple containing four integers representing the boundaries of the tissue grid.</dt><dd><ul class="simple">
<li><p>x_min (int): The minimum x - coordinate of the tissue grid.</p></li>
<li><p>x_max (int): The maximum x - coordinate of the tissue grid, adjusted to be divisible by the patch size.</p></li>
<li><p>y_min (int): The minimum y - coordinate of the tissue grid.</p></li>
<li><p>y_max (int): The maximum y - coordinate of the tissue grid, adjusted to be divisible by the patch size.</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="storm.pp.VisiumPreprocesser.concat_adata">
<span class="sig-name descname"><span class="pre">concat_adata</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.VisiumPreprocesser.concat_adata" title="Link to this definition"></a></dt>
<dd><p>Concatenate the average, interpolated, and zero - valued AnnData objects.</p>
<p>This method creates a zero - valued AnnData object for the remaining grid points,
concatenates the average, interpolated, and zero - valued AnnData objects,
and ensures that the indices of the final AnnData object match the tissue grid indices.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="storm.pp.VisiumPreprocesser.crop_img">
<span class="sig-name descname"><span class="pre">crop_img</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.VisiumPreprocesser.crop_img" title="Link to this definition"></a></dt>
<dd><p>Crop and pad the H&amp;E image.</p>
<p>This method calculates the padding needed for the H&amp;E image based on the tissue grid,
segments the tissue in the image, balances the white color,
pads the image with white space, and calculates color features before and after processing.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><dl class="simple">
<dt>A tuple containing six elements:</dt><dd><ul class="simple">
<li><p>raw_he_feature (np.ndarray): A 1D NumPy array representing the median color values of the tissue foreground
in the raw H&amp;E image. Each element corresponds to a color channel (e.g., RGB).</p></li>
<li><p>raw_bg_feature (np.ndarray): A 1D NumPy array representing the median color values of the tissue background
in the raw H&amp;E image. Each element corresponds to a color channel (e.g., RGB).</p></li>
<li><p>after_he_feature (np.ndarray): A 1D NumPy array representing the median color values of the tissue foreground
in the processed H&amp;E image. Each element corresponds to a color channel (e.g., RGB).</p></li>
<li><p>after_bg_feature (np.ndarray): A 1D NumPy array representing the median color values of the tissue background
in the processed H&amp;E image. Each element corresponds to a color channel (e.g., RGB).</p></li>
<li><p>raw_he (np.ndarray): A 3D NumPy array representing the raw H&amp;E image with shape (height, width, channels).</p></li>
<li><p>pd_he (np.ndarray): A 3D NumPy array representing the processed and padded H&amp;E image with shape (height, width, channels).</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="storm.pp.VisiumPreprocesser.expand_barcode">
<span class="sig-name descname"><span class="pre">expand_barcode</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.VisiumPreprocesser.expand_barcode" title="Link to this definition"></a></dt>
<dd><p>Expand barcode information and mark tissue grid labels.</p>
<p>This method adds center coordinates and a number column to the final_tpl DataFrame,
initializes two arrays for barcode and tissue labels,
marks barcode and tissue grid labels,
finds the nearest in - tissue point for each grid point,
expands the barcode labels, and creates a DataFrame for barcode information.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="storm.pp.VisiumPreprocesser.find_avg_grid">
<span class="sig-name descname"><span class="pre">find_avg_grid</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.VisiumPreprocesser.find_avg_grid" title="Link to this definition"></a></dt>
<dd><p>Find the average grid and create an AnnData object for it.</p>
<p>This method filters the tissue grid to include only points with valid barcodes and in - tissue status,
calculates the average gene expression for each barcode group,
and creates an AnnData object for the average grid.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="storm.pp.VisiumPreprocesser.generate_grid">
<span class="sig-name descname"><span class="pre">generate_grid</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.VisiumPreprocesser.generate_grid" title="Link to this definition"></a></dt>
<dd><p>Generate tissue grid.</p>
<p>This method calculates the tissue grid boundaries using the cal_cor method,
creates a meshgrid based on the boundaries and grid spacing,
and stores the resulting grid in a DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="storm.pp.VisiumPreprocesser.insert_grid">
<span class="sig-name descname"><span class="pre">insert_grid</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.VisiumPreprocesser.insert_grid" title="Link to this definition"></a></dt>
<dd><p>Interpolate gene expression for the non - average grid points.</p>
<p>This method filters the tissue grid to include only non - average grid points within the tissue,
uses nearest neighbor interpolation to estimate gene expression for these points,
and creates an AnnData object for the interpolated grid.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="storm.pp.VisiumPreprocesser.map_tissue">
<span class="sig-name descname"><span class="pre">map_tissue</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.VisiumPreprocesser.map_tissue" title="Link to this definition"></a></dt>
<dd><p>Map tissue information to the tissue grid.</p>
<p>This method expands barcode information, joins the barcode DataFrame to the tissue grid,
fills missing barcode values, and adds spot coordinates to the tissue grid.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="storm.pp.VisiumPreprocesser.mark_bc_label">
<span class="sig-name descname"><span class="pre">mark_bc_label</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">row</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.VisiumPreprocesser.mark_bc_label" title="Link to this definition"></a></dt>
<dd><p>Mark barcode labels in the array_forBarcode.</p>
<p>This method assigns a number to a specific position in the array_forBarcode based on the center coordinates of a row.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>row</strong> (<em>pandas.Series</em>) – A row from the DataFrame containing ‘center_x’ and ‘center_y’ columns.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="storm.pp.VisiumPreprocesser.mark_tissue_grid">
<span class="sig-name descname"><span class="pre">mark_tissue_grid</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">row</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.VisiumPreprocesser.mark_tissue_grid" title="Link to this definition"></a></dt>
<dd><p>Mark tissue grid labels.</p>
<p>This method attempts to assign the ‘in_tissue’ value of a row to a specific index in the tissue_grid DataFrame.
If the index does not exist, it adds a new row to the DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>row</strong> (<em>pandas.Series</em>) – A row from the DataFrame containing ‘center_x’ and ‘center_y’ columns.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="storm.pp.VisiumPreprocesser.mark_tissue_label">
<span class="sig-name descname"><span class="pre">mark_tissue_label</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">row</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.VisiumPreprocesser.mark_tissue_label" title="Link to this definition"></a></dt>
<dd><p>Mark tissue labels in the array_forIntissue.</p>
<p>This method attempts to assign the ‘in_tissue’ value of a row to a specific position in the array_forIntissue.
If an exception occurs, the method does nothing.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>row</strong> (<em>pandas.Series</em>) – A row from the DataFrame containing ‘tl_xn’ and ‘tl_yn’ columns.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="storm.pp.VisiumPreprocesser.process_adata">
<span class="sig-name descname"><span class="pre">process_adata</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.VisiumPreprocesser.process_adata" title="Link to this definition"></a></dt>
<dd><p>Process the AnnData object.</p>
<p>This method finds the average grid, interpolates gene expression for non - average grid points,
and concatenates the resulting AnnData objects.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">storm.VisiumReader</span> <span class="kn">import</span> <span class="n">VisiumReader</span>
<span class="kn">from</span> <span class="nn">storm.pp</span> <span class="kn">import</span> <span class="n">VisiumPreprocesser</span>
<span class="n">Reader</span><span class="o">=</span><span class="n">VisiumReader</span><span class="p">()</span>
<span class="n">Reader</span><span class="o">.</span><span class="n">read_all</span><span class="p">(</span><span class="n">folder_path</span><span class="o">=</span><span class="s2">&quot;../Visium_Human_Breast_Cancer&quot;</span><span class="p">,</span><span class="n">gene_token</span><span class="o">=</span><span class="s2">&quot;../gene_token_homologs.csv&quot;</span><span class="p">,</span><span class="n">method</span><span class="o">=</span><span class="s2">&quot;binary&quot;</span><span class="p">,</span><span class="n">key</span><span class="o">=</span><span class="s2">&quot;symbol&quot;</span><span class="p">)</span>
<span class="n">processer</span><span class="o">=</span><span class="n">VisiumPreprocesser</span><span class="p">(</span><span class="n">Reader</span><span class="p">,</span><span class="mi">224</span><span class="p">)</span>
<span class="n">processer</span><span class="o">.</span><span class="n">process_tpl</span><span class="p">()</span>
<span class="n">processer</span><span class="o">.</span><span class="n">round_spot</span><span class="p">()</span>
<span class="n">processer</span><span class="o">.</span><span class="n">generate_grid</span><span class="p">()</span>
<span class="n">processer</span><span class="o">.</span><span class="n">map_tissue</span><span class="p">()</span>
<span class="n">processer</span><span class="o">.</span><span class="n">process_adata</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">processer</span><span class="o">.</span><span class="n">fnl_adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>gene id matches:
human:15597 
mouse:16
</pre></div>
</div>
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>3839
</pre></div>
</div>
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 50686 × 15757
    var: &#39;gene_ids_x&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;symbol_x&#39;, &#39;gene_ids_y&#39;, &#39;symbol_y&#39;
</pre></div>
</div>
</div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="storm.pp.VisiumPreprocesser.process_img">
<span class="sig-name descname"><span class="pre">process_img</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.VisiumPreprocesser.process_img" title="Link to this definition"></a></dt>
<dd><p>Process the H&amp;E image.</p>
<p>This method processes the tissue position data, rounds the spot coordinates,
generates the tissue grid, maps tissue information, and crops the H&amp;E image.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><dl class="simple">
<dt>A tuple containing six elements returned by the crop_img method.</dt><dd><ul class="simple">
<li><p>raw_he_feature (np.ndarray): A 1D NumPy array representing the median color values of the tissue foreground
in the raw H&amp;E image. Each element corresponds to a color channel (e.g., RGB).</p></li>
<li><p>raw_bg_feature (np.ndarray): A 1D NumPy array representing the median color values of the tissue background
in the raw H&amp;E image. Each element corresponds to a color channel (e.g., RGB).</p></li>
<li><p>after_he_feature (np.ndarray): A 1D NumPy array representing the median color values of the tissue foreground
in the processed H&amp;E image. Each element corresponds to a color channel (e.g., RGB).</p></li>
<li><p>after_bg_feature (np.ndarray): A 1D NumPy array representing the median color values of the tissue background
in the processed H&amp;E image. Each element corresponds to a color channel (e.g., RGB).</p></li>
<li><p>raw_he (np.ndarray): A 3D NumPy array representing the raw H&amp;E image with shape (height, width, channels).</p></li>
<li><p>pd_he (np.ndarray): A 3D NumPy array representing the processed and padded H&amp;E image with shape (height, width, channels).</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>tuple</p>
</dd>
</dl>
<div class="jupyter_cell jupyter_container docutils container">
<div class="cell_input code_cell docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">storm.VisiumReader</span> <span class="kn">import</span> <span class="n">VisiumReader</span>
<span class="kn">from</span> <span class="nn">storm.pp</span> <span class="kn">import</span> <span class="n">VisiumPreprocesser</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="n">Reader</span><span class="o">=</span><span class="n">VisiumReader</span><span class="p">()</span>
<span class="n">Reader</span><span class="o">.</span><span class="n">read_all</span><span class="p">(</span><span class="n">folder_path</span><span class="o">=</span><span class="s2">&quot;../Visium_Human_Breast_Cancer&quot;</span><span class="p">,</span><span class="n">gene_token</span><span class="o">=</span><span class="s2">&quot;../gene_token_homologs.csv&quot;</span><span class="p">,</span><span class="n">method</span><span class="o">=</span><span class="s2">&quot;binary&quot;</span><span class="p">,</span><span class="n">key</span><span class="o">=</span><span class="s2">&quot;symbol&quot;</span><span class="p">)</span>

<span class="n">processer</span><span class="o">=</span><span class="n">VisiumPreprocesser</span><span class="p">(</span><span class="n">Reader</span><span class="p">,</span><span class="mi">224</span><span class="p">)</span>
<span class="n">_</span><span class="p">,</span><span class="n">_</span><span class="p">,</span><span class="n">_</span><span class="p">,</span><span class="n">_</span><span class="p">,</span><span class="n">raw_crop_he</span><span class="p">,</span><span class="n">crop_he</span><span class="o">=</span><span class="n">processer</span><span class="o">.</span><span class="n">process_img</span><span class="p">()</span>
<span class="n">fig</span><span class="p">,</span><span class="n">axs</span><span class="o">=</span><span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">raw_crop_he</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">crop_he</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">processer</span><span class="o">.</span><span class="n">tissue_position_list</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">),</span><span class="n">processer</span><span class="o">.</span><span class="n">final_tpl</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">),</span><span class="n">processer</span><span class="o">.</span><span class="n">tissue_grid</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>gene id matches:
human:15597 
mouse:16
</pre></div>
</div>
<div class="output stream highlight-none notranslate"><div class="highlight"><pre><span></span>                    in_tissue  spot_y  spot_x  pxl_y_fullres  pxl_x_fullres  \
barcode                                                                       
ACGCCTGACACGCGCT-1        1.0     0.0     0.0         4650.0        12203.0   
TACCGATCCAACACTT-1        1.0     1.0     1.0         4887.0        12338.0   
ATTAAAGCGGACGAGC-1        1.0     0.0     2.0         4651.0        12475.0   
GATAAGGGACGATTAG-1        1.0     1.0     3.0         4888.0        12609.0   
GTGCAAATCACCAATA-1        1.0     0.0     4.0         4653.0        12747.0   

                    pxl_y_hires  pxl_x_hires  \
barcode                                        
ACGCCTGACACGCGCT-1   223.708264   587.077837   
TACCGATCCAACACTT-1   235.110169   593.572593   
ATTAAAGCGGACGAGC-1   223.756373   600.163567   
GATAAGGGACGATTAG-1   235.158278   606.610214   
GTGCAAATCACCAATA-1   223.852592   613.249298   

                                            top_left_before top_left_after  
barcode                                                                     
ACGCCTGACACGCGCT-1  (582.8310287943084, 219.46145568230844)     (580, 216)  
TACCGATCCAACACTT-1  (589.3257848343085, 230.86336073030844)     (588, 228)  
ATTAAAGCGGACGAGC-1  (595.9167594823084, 219.50956498630842)     (592, 216)  
GATAAGGGACGATTAG-1  (602.3634062183085, 230.91147003430842)     (600, 228)  
GTGCAAATCACCAATA-1  (609.0024901703084, 219.60578359430843)     (608, 216)                       in_tissue  spot_y  spot_x  tl_x  tl_y  center_x  center_y  \
barcode                                                                         
ACGCCTGACACGCGCT-1        1.0     0.0     0.0     4     0        12        -8   
TACCGATCCAACACTT-1        1.0     1.0     1.0    12    12        20         4   
ATTAAAGCGGACGAGC-1        1.0     0.0     2.0    16     0        24        -8   
GATAAGGGACGATTAG-1        1.0     1.0     3.0    24    12        32         4   
GTGCAAATCACCAATA-1        1.0     0.0     4.0    32     0        40        -8   

                    num  
barcode                  
ACGCCTGACACGCGCT-1    1  
TACCGATCCAACACTT-1    2  
ATTAAAGCGGACGAGC-1    3  
GATAAGGGACGATTAG-1    4  
GTGCAAATCACCAATA-1    5                 tl_xn  tl_yn  in_tissue      position          barcode_55  \
index                                                                     
s_004_0_0-n     0.0    0.0        1.0   s_004_0_0-n                   0   
s_004_4_0-n     4.0    0.0        1.0   s_004_4_0-n                   0   
s_004_8_0-n     8.0    0.0        1.0   s_004_8_0-n                   0   
s_004_12_0-n   12.0    0.0        1.0  s_004_12_0-n                   0   
s_004_16_0-n   16.0    0.0        1.0  s_004_16_0-n  TACCGATCCAACACTT-1   

              spot_x  spot_y  
index                         
s_004_0_0-n      0.0     0.0  
s_004_4_0-n      1.0     0.0  
s_004_8_0-n      2.0     0.0  
s_004_12_0-n     3.0     0.0  
s_004_16_0-n     4.0     0.0  
</pre></div>
</div>
<img alt="_images/pp_1_4.png" src="_images/pp_1_4.png" />
</div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="storm.pp.VisiumPreprocesser.process_tpl">
<span class="sig-name descname"><span class="pre">process_tpl</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.VisiumPreprocesser.process_tpl" title="Link to this definition"></a></dt>
<dd><p>Process the tissue position data in a CSV - like DataFrame.</p>
<p>This method modifies the internal DataFrame ‘tissue_position_list’ directly.
It sets the index to ‘barcode’, removes rows until the first barcode starts with ‘0’ or ‘1’,
converts all columns to float type, scales certain pixel columns, renames columns for clarity,
and adds two new columns initialized as None.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>None</strong> (<em>This is an instance method and does not require any arguments.</em>)</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="storm.pp.VisiumPreprocesser.round_spot">
<span class="sig-name descname"><span class="pre">round_spot</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.VisiumPreprocesser.round_spot" title="Link to this definition"></a></dt>
<dd><p>Round the coordinates of spots to grid lines.</p>
<p>This method calculates the radius of a circle based on tissue scale factor and spot diameter.
It rounds the top - left corner coordinates of each tissue position to the nearest grid line (with 4 as the grid size),
crops the H&amp;E image using the bounding box of the rounded tissue positions with an additional spot area,
and finally transforms the tissue position list by normalizing the rounded top - left coordinates.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="storm.pp.VisiumPreprocesser.save">
<span class="sig-name descname"><span class="pre">save</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">final_grid_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">final_h5ad_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">final_png_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">final_color_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.VisiumPreprocesser.save" title="Link to this definition"></a></dt>
<dd><p>Save the processed data to files.</p>
<p>This method processes the H&amp;E image, saves color features, the H&amp;E image,
the tissue grid, and the final AnnData object to files if the corresponding file paths are provided.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>final_grid_path</strong> (<em>str</em><em>, </em><em>optional</em>) – The path to save the tissue grid as a CSV file. Defaults to None.</p></li>
<li><p><strong>final_h5ad_path</strong> (<em>str</em><em>, </em><em>optional</em>) – The path to save the final AnnData object. Defaults to None.</p></li>
<li><p><strong>final_png_path</strong> (<em>str</em><em>, </em><em>optional</em>) – The path to save the processed and raw H&amp;E images as PNG files. Defaults to None.</p></li>
<li><p><strong>final_color_path</strong> (<em>str</em><em>, </em><em>optional</em>) – The path to save the color features. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="storm.pp.downsample_svs">
<span class="sig-prename descclassname"><span class="pre">storm.pp.</span></span><span class="sig-name descname"><span class="pre">downsample_svs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_svs_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_tiff_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_mpp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.downsample_svs" title="Link to this definition"></a></dt>
<dd><p>Downsample an SVS whole slide image to specified target microns per pixel (MPP) and save as TIFF</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_svs_path</strong> (<em>str</em>) – Path to input SVS file</p></li>
<li><p><strong>output_tiff_path</strong> (<em>str</em>) – Path to save output TIFF file</p></li>
<li><p><strong>target_mpp</strong> – Target resolution in microns per pixel (default: 0.5)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="storm.pp.exc_he">
<span class="sig-prename descclassname"><span class="pre">storm.pp.</span></span><span class="sig-name descname"><span class="pre">exc_he</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">img</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stain_matrix</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.exc_he" title="Link to this definition"></a></dt>
<dd><p>Extract Hematoxylin and Eosin channels from an image.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>img</strong> (<em>np.ndarray</em>) – The input RGB image.</p></li>
<li><p><strong>stain_matrix</strong> (<em>np.ndarray</em>) – The stain matrix.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A tuple containing the Hematoxylin and Eosin channel images.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="storm.pp.exc_tissue">
<span class="sig-prename descclassname"><span class="pre">storm.pp.</span></span><span class="sig-name descname"><span class="pre">exc_tissue</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">img</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'lumin'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">l_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.exc_tissue" title="Link to this definition"></a></dt>
<dd><p>Generate a tissue mask from an RGB image using either luminance thresholding or Otsu’s method.</p>
<p>This function takes an RGB image and creates a binary mask to identify the tissue region within the image.
It supports two methods: luminance thresholding and Otsu’s method for automatic threshold selection.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>img</strong> (<em>ndarray</em>) – An RGB image represented as a NumPy array with shape (height, width, 3).</p></li>
<li><p><strong>method</strong> (<em>str</em><em>, </em><em>optional</em>) – The method to use for generating the tissue mask.
Options are ‘lumin’ for luminance thresholding and ‘otsu’ for Otsu’s method. Defaults to ‘lumin’.</p></li>
<li><p><strong>l_threshold</strong> (<em>float</em><em>, </em><em>optional</em>) – The luminance threshold value used when <cite>method</cite> is set to ‘lumin’.
Values range from 0 to 1. Defaults to 0.8.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>A binary mask representing the detected tissue region in the input image.</dt><dd><p>It has the same height and width as the input image.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>ndarray</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the <cite>method</cite> argument is not ‘lumin’ or ‘otsu’.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="storm.pp.extract_tiles">
<span class="sig-prename descclassname"><span class="pre">storm.pp.</span></span><span class="sig-name descname"><span class="pre">extract_tiles</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tiff_folder</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">h5_folder</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_folder</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.extract_tiles" title="Link to this definition"></a></dt>
<dd><p>Extract image tiles from TIFF images using coordinates stored in H5 files</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tiff_folder</strong> (<em>str</em>) – Path to folder containing TIFF image files</p></li>
<li><p><strong>h5_folder</strong> (<em>str</em>) – Path to folder containing H5 files with coordinate data</p></li>
<li><p><strong>output_folder</strong> – Path to folder where extracted tiles will be saved</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="storm.pp.process_folder">
<span class="sig-prename descclassname"><span class="pre">storm.pp.</span></span><span class="sig-name descname"><span class="pre">process_folder</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_mpp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.process_folder" title="Link to this definition"></a></dt>
<dd><p>Batch process SVS files in a directory to downsample and convert to TIFF format</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>input_dir</strong> (<em>str</em>) – Path to directory containing input SVS files</p></li>
<li><p><strong>output_dir</strong> (<em>str</em>) – Path to directory where downsampled TIFF files will be saved</p></li>
<li><p><strong>target_mpp</strong> (<em>float</em>) – Target microns per pixel for downsampling (default: 0.5)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="storm.pp.white_balance_using_white_point">
<span class="sig-prename descclassname"><span class="pre">storm.pp.</span></span><span class="sig-name descname"><span class="pre">white_balance_using_white_point</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">img</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#storm.pp.white_balance_using_white_point" title="Link to this definition"></a></dt>
<dd><p>Perform white balance on an RGB image using the white point method.</p>
<p>This function adjusts the color balance of an RGB image by calculating the white point
from the masked region of the image. It then applies gain factors to each color channel
to achieve a more natural color representation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>img</strong> (<em>ndarray</em>) – An RGB image represented as a NumPy array with shape (height, width, 3).
The pixel values should be in the range [0, 255].</p></li>
<li><p><strong>mask</strong> (<em>ndarray</em>) – A binary mask with the same height and width as the input image.
The mask is used to select the region of interest for calculating the white point.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>The white - balanced RGB image with the same shape as the input image.</dt><dd><p>The pixel values are in the range [0, 255].</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>ndarray</p>
</dd>
</dl>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="VisiumReader.html" class="btn btn-neutral float-left" title="VisiumReader" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="gt.html" class="btn btn-neutral float-right" title="gt" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Zongxu Zhang.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>